### ISSUE TITLE
ChartViewDelegate Function func chartValueSelected(_ chartView: ChartViewBase, entry: ChartDataEntry, highlight: Highlight) is not being called for every tap/touch inside the ChartView

### ISSUE THREAD

**Issue by kiranGrg** [danielgindi/Charts#4719]
* [ ] I've read, understood, and done my best to follow the [*CONTRIBUTING guidelines](https://github.com/jjatie/Charts/blob/master/CONTRIBUTING.md).

## What did you do?

I upgraded the Charts(iOS) version to 4.0.1 from 3.6.0 recently. Fixed all the changes in the codebase. Everything seems to be working except for one behavior on the chart. 

## What did you expect to happen?

Enabled the user interaction on the chart. Added the ChartViewDelegate, and implemented the optional delegate function
`public func chartValueSelected(_ chartView: ChartViewBase, entry: ChartDataEntry, highlight: Highlight)`

## What happened instead?

However, apparently, this delegate function `public func chartValueSelected(_ chartView: ChartViewBase, entry: ChartDataEntry, highlight: Highlight)` is not being triggered every time I tap inside the chart view (to be specific I was using LineChartView), instead only for some places in the chart view seems to be responding to the tap/touch.

I am pretty, sure the delegate function used to be called with failing in the Charts 3.6.0.

## Additional Findings:
I tried the official demo version for iOS source codes for both 3.6.0 and 4.0.1, as i couldn't get this issue fixed. Looks like, the demo itself is exhibiting this problem, as I could see the delegate function being called with every tap/touche in 3.6.0 but not in the demo version 4.0.1

## Charts Environment

**Charts version/Branch/Commit Number: 4.0.1**
**Xcode version: 12.5.1**
**Swift version: 5.4.2**
**Platform(s) running Charts: iOS device/simulators**
**macOS version running Xcode: 11.5.2**

**Comment by esma01** [comment]
I see the same issue. The sample project is also not working! Only the middle and last data points can be selected.

**Comment by SergejLogis** [comment]
Same here, 100% accurate description of an issue and same observations (Charts demo app)...

**Comment by PhillipMaizza** [comment]
I'm experiencing the same issue with line charts.

**Comment by esma01** [comment]
@PhillipMaizza @SergejLogis Downgrading the charts to 3.6.0 worked for me!

**Comment by SergejLogis** [comment]
Yeah, that helped me as well. Only I didn't downgrade the entire library to 3.6.0, but instead created a custom `LineChartDataSet` with a method overridden with code from 3.6.0:

```
class CustomLineChartDataSet: LineChartDataSet {

    override func entriesForXValue(_ xValue: Double) -> [ChartDataEntry] {
        var entries = [ChartDataEntry]()

        var low = startIndex
        var high = endIndex - 1

        while low <= high {
            var mid = (high + low) / 2
            var entry = self[mid]

            // if we have a match
            if xValue == entry.x {
                while mid > 0 && self[mid - 1].x == xValue {
                    mid -= 1
                }

                high = endIndex

                // loop over all "equal" entries
                while mid < high {
                    entry = self[mid]
                    if entry.x == xValue {
                        entries.append(entry)
                    } else {
                        break
                    }

                    mid += 1
                }

                break

            } else {
                if xValue > entry.x {
                    low = mid + 1
                } else {
                    high = mid - 1
                }
            }
        }

        return entries
    }

}
```

**Comment by PhillipMaizza** [comment]
I downgraded as well. Nice workaround though, @SergejLogis !

**Comment by kcome** [comment]
@liuxuan30 @jjatie  Thanks for having a look at [this PR](https://github.com/danielgindi/Charts/pull/4721) that fixes this issue.

**Comment by SergejLogis** [comment]
@kcome I've tried your solution and it works like charm! 🥰 Much better than the one back-ported from 3.6.0. 🚀🙇‍♂️

**Comment by kcome** [comment]
Hi @hrvoje0099 , I did another test and found the issue you described. And actually I just fixed it in another problem I encountered so today I merged the fix. If that suites you could you have a try? (this is the [commit](https://github.com/danielgindi/Charts/pull/4721/commits/85cc87d30d6ac68181380272eea1e4e9dc5d8147))
(sorry for late reply)

**Comment by faaizdaglawala** [comment]
> > @kcome I am facing crash in this solution. after zooming the chart and then reloading the chart data crashes the app.
> 
> @faaizdaglawala I tested the demo app on simulator (xcode 13.3, iOS 15), I don't find crash after zooming and reloading chart data. Could you give an example project and specific steps of reproducing it? Also could you give specific about "this solution"? Is it some comment in this issue discussion or a branch from some repo?

I had one chart view in which there are 2 lines in and then I reload that chart and add one line and one bar chart (CombinedChartView), now the scenario is when I zoom the chart when it is having one line and one bar chart (CombinedChartView), it works and then when I reload that chart to 2 lines then it crashes in subscript method saying index out of range (Will share video shortly)

**Comment by kcome** [comment]
@faaizdaglawala what you described in the comment and screenshots is probably another issue which is not related to this issue and the fix. I could reproduce a crash which is similar to your description, however the crash line is not the same as yours. Could you have a look at this #4798?

**Comment by faaizdaglawala** [comment]
@kcome understood your point but when I remove your changes, it starts working.

**Comment by faaizdaglawala** [comment]
@pavel-zlotarenchuk I downgraded the charts to version 3.6.0.

**Comment by dudenamedjune** [comment]
Think I have a proper fix for this 😅 ... was a debugging rodeo 

https://github.com/danielgindi/Charts/pull/4817

**Comment by PorterHoskins** [comment]
+1

**Comment by PorterHoskins** [comment]
Downgrading is not an option for people using SPM. It would be really nice to get this fix released.

**Comment by zygoat** [comment]
@PorterHoskins What prevents you from downgrading?

I import Charts from SPM; albeit my dependencies are configured in Xcode's `pbxproj`, but the equivalent in `Packages.swift` would be `.upToNextMajor(from: "3.0.0")` in the dependency spec.

**Comment by PorterHoskins** [comment]
Ok that worked... I was confused by the SPM support documentation in the 4.0 release. Sorry

**Comment by pmairoldi** [comment]
Closed by https://github.com/danielgindi/Charts/pull/4721
